packages/@atjson/document/src/annotation.ts

```diff
@@ -1,18 +1,158 @@
-import { Display } from './schema';
 
-export default interface Annotation {
-  id?: string | number;
-  type: string;
-  display?: Display;
+import { Attributes, toJSON, unprefix } from './attributes';
+import Change, { AdjacentBoundaryBehaviour, Deletion, Insertion } from './change';
+
+export default abstract class Annotation {
+  static vendorPrefix: string;
+  static type: string;
+
+  static hydrate<T extends Annotation>(
+    this: {
+      vendorPrefix: string;
+      new(attributes: { start: number, end: number, attributes: Attributes }): T;
+    },
+    attrs: { start: number, end: number, attributes: Attributes }
+  ) {
+    return new this({
+      start: attrs.start,
+      end: attrs.end,
+      attributes: unprefix(this.vendorPrefix, attrs.attributes) as Attributes
+    });
+  }
+
+  readonly type: string;
+  abstract rank: number;
   start: number;
   end: number;
+  attributes: Attributes;
+
+  constructor(attrs: { start: number, end: number, attributes: Attributes }) {
+    let AnnotationClass = this.constructor as typeof Annotation;
+    this.type = AnnotationClass.type;
+    this.start = attrs.start;
+    this.end = attrs.end;
+
+    this.attributes = attrs.attributes;
+  }
+
+  /**
+   * nb. Currently, changes are applied directly to the document.
+   *     In the future, we want to return a set of changes that
+   *     are applied to the document including annotations.
+   */
+  handleChange(change: Change) {
```

Comment:
Does it make sense to also abstract adding/removing/changing annotations as `Change`s? High-level, that feels the right approach, but it might be impractical at the moment.